### ISSUE TITLE
compile for windows fail after the last commit in aloptional.h

### ISSUE THREAD

**Issue by SiminBadri** [kcat/openal-soft#603]
Hi,

I cloned openal-soft last week and could build for different platforms successfully. However, today it fails in building for windows with the following error: 

```
aloptional.h(223): error C2512: 'al::optional_storage<T,false,false,false,false>::optional_storage': no appropriate default constructor available
        with
        [
            T=std::wstring
        ]
```

About 9 hours ago, I tried to build for windows but it fails with a different error. After @kcat last commit, that error is fixed but the above error occurs.

**Comment by kcat** [comment]
What version of MSVC are you using? I've been trying to get it to build the code since it is helpful, and everything looks correct and both GCC and Clang handle it without problems, so I can only assume it's an MSVC issue. I can revert the change, but preferably only the versions that have trouble with the new code so newer MSVC versions can get its benefits, if possible.

**Comment by SiminBadri** [comment]
I use vs 2019 with MSVC Version 14.20.27508.1. I changed some files and now it successfully builds. I share the changes with you, however, I am not sure to what extent it complies with the purpose you follow.

In file ```aloptional.h```, I changed the struct and add a default constructor:

``` cpp
optional_storage() = default;
```

``` cpp
/* Completely non-trivial. */
template<typename T>
struct optional_storage<T, false, false, false, false> : public optstore_helper<T> {
    using optstore_helper<T>::optstore_helper;
    optional_storage() = default;
    optional_storage(const optional_storage &rhs) NOEXCEPT_AS(this->construct(rhs.mValue))
    { if(rhs.mHasValue) this->construct(rhs.mValue); }
    optional_storage(optional_storage&& rhs) NOEXCEPT_AS(this->construct(std::move(rhs.mValue)))
    { if(rhs.mHasValue) this->construct(std::move(rhs.mValue)); }
    optional_storage& operator=(const optional_storage &rhs) NOEXCEPT_AS(this->assign(rhs))
    { this->assign(rhs); return *this; }
    optional_storage& operator=(optional_storage&& rhs) NOEXCEPT_AS(this->assign(std::move(rhs)))
    { this->assign(std::move(rhs)); return *this; }
};
```

Also, I added the following lines after the above struct definition:

``` cpp
template<typename T>
struct optional_storage<T, true, true, true, true> : public optstore_helper<T> {
    using optstore_helper<T>::optstore_helper;
    optional_storage() = default;
    optional_storage(const optional_storage &rhs) NOEXCEPT_AS(this->construct(rhs.mValue))
    { if(rhs.mHasValue) this->construct(rhs.mValue); }
    optional_storage(optional_storage&& rhs) NOEXCEPT_AS(this->construct(std::move(rhs.mValue)))
    { if(rhs.mHasValue) this->construct(std::move(rhs.mValue)); }
    optional_storage& operator=(const optional_storage &rhs) NOEXCEPT_AS(this->assign(rhs))
    { this->assign(rhs); return *this; }
    optional_storage& operator=(optional_storage&& rhs) NOEXCEPT_AS(this->assign(std::move(rhs)))
    { this->assign(std::move(rhs)); return *this; }
};
```

With the above changes, the error in ```aloptional.h``` is fixed, however, another error occurs in the file ``` event.cpp``` which stems from the file ```almalloc.h```.  The error is:

```
almalloc.h(109): error C2398: Element '1': conversion from '_Ty' to 'uint' requires a narrowing conversion
        with
        [
            _Ty=<unnamed-enum-EventType_KillThread>
        ]
```

This is because ```EventType_KillThread``` is an enum (with default type ```int```), however, the type in ```AsyncEvent``` which uses this enum is defined as ```uint``` which may cause an error in the assignment. If I am right, the enum used in ```async_event.h``` has only positive values, so it's safe to define this enum as `uint`. Therefore, I changed the enum in the file ```async_event.h``` as the following line:

``` cpp
enum : uint{
    /* End event thread processing. */
    EventType_KillThread = 0,

    /* User event types. */
    EventType_SourceStateChange = 1<<0,
    EventType_BufferCompleted   = 1<<1,
    EventType_Disconnected      = 1<<2,

    /* Internal events. */
    EventType_ReleaseEffectState = 65536,
};
```

After the above changes, the library was built successfully for the windows platform.

**Comment by kcat** [comment]
I made the explicit constructor change with commit 6a3c4c09f838b0e9156b3f4c223281a3ed43817b, and the EventType enum fix with commit 92b65fa15fa8d7ca064ef6178fc749b6a465ad5b, and it seems to be building now on both GitHub and AppVeyor CI. I didn't do the second change for `optional_storage<T, true, true, true, true>` since that would prevent `al::optional` from being trivially copyable and movable when holding a trivially copyable or movable type, which was the intent of the initial change because it allows the compiler to better optimize it.

Let me know if you're still getting problems, and I'll do what I can to work around it.

**Comment by SiminBadri** [comment]
I cloned your last commit and now it successfully builds for windows x86 and x64.

Thanks
